R/expect-shape.R

```diff
@@ -15,42 +15,53 @@
 #' @export
 #' @examples
 expect_shape = function(object, ..., length, nrow, ncol, dim) {
-  # absent, not present, due to '!' operator precedence
-  n_absent <- missing(length) + missing(nrow) + missing(ncol) + missing(dim)
-  if (n_absent != 3L) {
-    cli::cli_abort(
-      "Exactly one of {.arg length}, {.arg nrow}, {.arg ncol}, or {.arg dim} must be provided."
-    )
-  }
+  rlang::check_exclusive(length, nrow, ncol, dim)
 
   # Re-use expect_length() to ensure they stay in sync.
   if (!missing(length)) {
     return(expect_length(object, length))
   }
 
+  # need base:: qualification or we might trigger an error for missing(length)
+  length <- base::length
+
   act <- quasi_label(enquo(object), arg = "object")
-  # need base:: qualification or we might trigger an error for missing(dim)
   dim_object <- base::dim(object)
 
+  expect(
+    !is.null(dim_object),
+    sprintf("%s has no dimensions.", act$lab)
+  )
+
   if (!missing(nrow)) {
     act$nrow <- dim_object[1L]
 
     expect(
-      act$nrow == nrow,
+      identical(as.integer(act$nrow), as.integer(nrow)),
       sprintf("%s has %i rows, not %i.", act$lab, act$nrow, nrow)
     )
   } else if (!missing(ncol)) {
+    expect(
+      length(dim_object) >= 2L,
+      sprintf("%s has only one dimension.", act$lab)
+    )
+
     act$ncol <- dim_object[2L]
 
     expect(
-      act$ncol == ncol,
+      identical(as.integer(act$ncol), as.integer(ncol)),
       sprintf("%s has %i columns, not %i.", act$lab, act$ncol, ncol)
     )
   } else { # !missing(dim)
     act$dim <- dim_object
 
     expect(
-      isTRUE(all.equal(act$dim, dim)),
+      length(act$dim) == length(dim),
```

Comment:
There is some principle that I only vaguely remember and should have written down, but I think you want to ensure that there's only one path that will somewhere yield a `success()`. (Otherwise if this `expect_shape()` succeeds, two successes will be recorded by the reporter).

So I think you should write this as `if (length(act$dim) != length(dim)) fail(sprintf(...))`

Reply:
I almost wonder if using `expect()` is actually an anti-pattern because it makes it so hard to generate the correct number of successes.

Reply:
Glancing, `fail()` just wraps `expect()`, though, so won't it have the same problem? Or should it just be an early return `if (!...) return(fail(...))`?

Reply:
Relatedly, is there a clean way to check that there's only one success? I'm thinking with whatever the fix is, it would be good to add a few extra tests that say "succeeds once, and only once" (which would fail on the current commit)

Reply:
Further aside about this hidden rake:

4bc32fca30403d406b46aa2bba313b6cf6d3daa0 found that, AIUI, because `expect()` _does_ pass for one of the two invocations, `expect_success()` was reporting "good to go" even though the latter `expect()` failed.

Reply:
I think you can assume that `fail()` will always throw, so you can treat it like an early return (but you don't need the `return()`).

I'm not sure it's worth testing that an expectation returns only 1 success, because you'd need to test it for every possible code path. (And I don't believe ensuring a single success is sufficiently valuable to add that much clutter to all of the other tests)

Reply:
But maybe `expect_success()` should be "expect exactly one success and no failures". I think that would help eliminate the rake you discovered.

Reply:
Punting to #2110